### PR TITLE
vendor: switch to using go modules to manage dependencies

### PR DESCRIPTION
**Pull request by staebler**
https://issues.redhat.com/browse/CO-837

### CODE REVIEW

**Comment by staebler** [comment]
This is the list of projects that we were using before that have either been changed or removed as part of switching to go modules.

```
{
  "name": "cloud.google.com/go",
  "old": "v0.30.0",
  "new": "v0.49.0"
}
{
  "name": "github.com/Azure/go-autorest",
  "old": "v13.4.0",
  "new": "v13.0.0"
}
{
  "name": "github.com/MakeNowJust/heredoc",
  "old": "e9091a26100e",
  "new": "v0.0.0-20170808103936-bb23615498cd"
}
{
  "name": "github.com/NYTimes/gziphandler",
  "old": "v1.0.1",
  "new": "v1.1.1"
}
{
  "name": "github.com/PuerkitoBio/purell",
  "old": "v1.1.0",
  "new": "v1.1.1"
}
{
  "name": "github.com/beorn7/perks",
  "old": "3a771d992973",
  "new": "v1.0.1"
}
{
  "name": "github.com/blang/semver",
  "old": "v3.6.1",
  "new": "v3.5.1"
}
{
  "name": "github.com/certifi/gocertifi",
  "old": "2020.01.04",
  "new": "v0.0.0-20190506164543-d2eda7129713"
}
{
  "name": "github.com/chai2010/gettext-go",
  "old": "bf70f2a70fb1",
  "new": "v0.0.0-20160711120539-c6fed771bfd5"
}
{
  "name": "github.com/coreos/etcd",
  "old": "v3.3.9",
  "new": "v3.3.18"
}
{
  "name": "github.com/coreos/go-systemd",
  "old": "v17",
  "new": "v22.0.0"
}
{
  "name": "github.com/docker/docker",
  "old": "a9fbbdc8dd87",
  "new": "v0.7.3-0.20190410184157-6d18c6a06295"
}
{
  "name": "github.com/emicklei/go-restful",
  "old": "v2.8.0",
  "new": "v2.10.0"
}
{
  "name": "github.com/evanphx/json-patch",
  "old": "v4.2.0",
  "new": "v4.5.0"
}
{
  "name": "github.com/ghodss/yaml",
  "old": "v1.0.0",
  "new": "v1.0.1-0.20190212211648-25d852aebe32"
}
{
  "name": "github.com/go-openapi/jsonpointer",
  "old": "v0.17.0",
  "new": "v0.19.3"
}
{
  "name": "github.com/go-openapi/jsonreference",
  "old": "v0.17.0",
  "new": "v0.19.2"
}
{
  "name": "github.com/go-openapi/spec",
  "old": "v0.17.0",
  "new": "v0.19.3"
}
{
  "name": "github.com/go-openapi/swag",
  "old": "v0.17.0",
  "new": "v0.19.5"
}
{
  "name": "github.com/gogo/protobuf",
  "old": "v1.1.1",
  "new": "v1.3.1"
}
{
  "name": "github.com/golang/groupcache",
  "old": "6f2cf27854a4",
  "new": "v0.0.0-20191027212112-611e8accdfc9"
}
{
  "name": "github.com/golang/mock",
  "old": "v1.2.0",
  "new": "v1.3.1"
}
{
  "name": "github.com/golang/protobuf",
  "old": "v1.3.1",
  "new": "v1.3.2"
}
{
  "name": "github.com/google/gofuzz",
  "old": "24818f796faf",
  "new": "v1.1.0"
}
{
  "name": "github.com/google/uuid",
  "old": "v1.0.0",
  "new": "v1.1.1"
}
{
  "name": "github.com/googleapis/gnostic",
  "old": "v0.2.0",
  "new": "v0.3.1"
}
{
  "name": "github.com/hashicorp/golang-lru",
  "old": "v0.5.0",
  "new": "v0.5.3"
}
{
  "name": "github.com/imdario/mergo",
  "old": "v0.3.6",
  "new": "v0.3.8"
}
{
  "name": "github.com/json-iterator/go",
  "old": "v1.1.5",
  "new": "v1.1.8"
}
{
  "name": "github.com/konsorten/go-windows-terminal-sequences",
  "old": "v1.0.1",
  "new": "v1.0.2"
}
{
  "name": "github.com/mailru/easyjson",
  "old": "60711f1a8329",
  "new": "v0.0.0-20190626092158-b2ccc519800e"
}
{
  "name": "github.com/mattn/go-colorable",
  "old": "v0.1.0",
  "new": "v0.1.4"
}
{
  "name": "github.com/mattn/go-isatty",
  "old": "v0.0.4",
  "new": "v0.0.10"
}
{
  "name": "github.com/miekg/dns",
  "old": "v1.1.8",
  "new": "v1.1.15"
}
{
  "name": "github.com/onsi/ginkgo",
  "old": "v1.6.0",
  "new": "v1.11.0"
}
{
  "name": "github.com/onsi/gomega",
  "old": "v1.4.2",
  "new": "v1.8.1"
}
{
  "name": "github.com/openshift/cloud-credential-operator",
  "old": "9cc1c1abf898",
  "new": "v0.0.0-20200316201045-d10080b52c9e"
}
{
  "name": "github.com/openshift/cluster-api",
  "old": "046d74a3bd91",
  "new": "v0.0.0-20191129101638-b09907ac6668"
}
{
  "name": "github.com/openshift/cluster-api-provider-gcp",
  "old": "093a944a412b",
  "new": "v0.0.1-0.20200120152131-1b09fd9e7156"
}
{
  "name": "github.com/openshift/cluster-autoscaler-operator",
  "old": "19de17fd6e66",
  "new": "v0.0.0-20190521201101-62768a6ba480"
}
{
  "name": "github.com/openshift/machine-api-operator",
  "old": "94398d11715b",
  "new": "v0.2.1-0.20200310180732-c63fa2b143f0"
}
{
  "name": "github.com/petar/GoLLRB",
  "old": "33fb24c13b99",
  "new": ""
}
{
  "name": "github.com/peterbourgon/diskv",
  "old": "v3.0.0",
  "new": "v2.0.1"
}
{
  "name": "github.com/pkg/errors",
  "old": "v0.8.0",
  "new": "v0.9.1"
}
{
  "name": "github.com/pkg/profile",
  "old": "v1.4.0",
  "new": "v1.3.0"
}
{
  "name": "github.com/prometheus/client_model",
  "old": "5c3871d89910",
  "new": "v0.0.0-20191202183732-d1d2010b5bee"
}
{
  "name": "github.com/prometheus/common",
  "old": "c7de2306084e",
  "new": "v0.6.0"
}
{
  "name": "github.com/prometheus/procfs",
  "old": "185b4288413d",
  "new": "v0.0.8"
}
{
  "name": "github.com/russross/blackfriday",
  "old": "v1.4",
  "new": "v1.5.2"
}
{
  "name": "github.com/spf13/cobra",
  "old": "v0.0.3",
  "new": "v0.0.5"
}
{
  "name": "github.com/spf13/pflag",
  "old": "v1.0.3",
  "new": "v1.0.5"
}
{
  "name": "github.com/vmware/govmomi",
  "old": "8ddca4d4f357",
  "new": "v0.22.2"
}
{
  "name": "go.opencensus.io",
  "old": "aab39bd6a98b",
  "new": "v0.22.2"
}
{
  "name": "golang.org/x/crypto",
  "old": "0c41d7ab0a0e",
  "new": "v0.0.0-20191206172530-e9b2fee46413"
}
{
  "name": "golang.org/x/net",
  "old": "49bb7cea24b1",
  "new": "v0.0.0-20200226121028-0de0cce0169b"
}
{
  "name": "golang.org/x/oauth2",
  "old": "c57b0facaced",
  "new": "v0.0.0-20191202225959-858c2ad4c8b6"
}
{
  "name": "golang.org/x/sys",
  "old": "fa43e7bc11ba",
  "new": "v0.0.0-20191206220618-eeba5f6aabab"
}
{
  "name": "golang.org/x/text",
  "old": "v0.3.0",
  "new": "v0.3.2"
}
{
  "name": "golang.org/x/time",
  "old": "fbb02b2291d2",
  "new": "v0.0.0-20191024005414-555d28b269f0"
}
{
  "name": "golang.org/x/tools",
  "old": "4fd33079060a",
  "new": "v0.0.0-20200331202046-9d5940d49312"
}
{
  "name": "gomodules.xyz/jsonpatch",
  "old": "v2.0.0",
  "new": ""
}
{
  "name": "google.golang.org/api",
  "old": "v0.3.2",
  "new": "v0.13.0"
}
{
  "name": "google.golang.org/appengine",
  "old": "v1.2.0",
  "new": "v1.6.5"
}
{
  "name": "google.golang.org/genproto",
  "old": "af9cb2a35e7f",
  "new": "v0.0.0-20200122232147-0452cf42e150"
}
{
  "name": "google.golang.org/grpc",
  "old": "v1.15.0",
  "new": "v1.25.1"
}
{
  "name": "gopkg.in/AlecAivazis/survey.v1",
  "old": "v1.8.7",
  "new": "v1.8.9-0.20200217094205-6773bdf39b7f"
}
{
  "name": "gopkg.in/ini.v1",
  "old": "v1.42.0",
  "new": "v1.51.0"
}
{
  "name": "gopkg.in/natefinch/lumberjack.v2",
  "old": "v2.1",
  "new": "v2.0.0"
}
{
  "name": "gopkg.in/yaml.v2",
  "old": "v2.2.1",
  "new": "v2.2.8"
}
{
  "name": "k8s.io/cloud-provider",
  "old": "9775d7be8494",
  "new": "v0.0.0-20190409023720-1bc0c81fa51d"
}
{
  "name": "k8s.io/component-base",
  "old": "327675bd8ec3",
  "new": "v0.17.1"
}
{
  "name": "k8s.io/gengo",
  "old": "e17681d19d3a",
  "new": "v0.0.0-20190907103519-ebc107f98eab"
}
{
  "name": "k8s.io/klog",
  "old": "v0.1.0",
  "new": "v1.0.0"
}
{
  "name": "sigs.k8s.io/cluster-api-provider-azure",
  "old": "832a43a30d7f",
  "new": "v0.1.0-alpha.3.0.20200120114645-8a9592f1f87b"
}
{
  "name": "sigs.k8s.io/testing_frameworks",
  "old": "v0.1.1",
  "new": "v0.1.2-0.20190130140139-57f07443c2d4"
}
```

**Comment by staebler** [comment]
I need to upgrade the go version used in the Docker files to go1.13.

**Comment by staebler** [comment]
https://github.com/openshift/hive/pull/918

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	golang.org/x/lint v0.0.0-20200302205851-738671d3881b // indirect
+	golang.org/x/net v0.0.0-20200226121028-0de0cce0169b
+	golang.org/x/oauth2 v0.0.0-20191202225959-858c2ad4c8b6
+	golang.org/x/tools v0.0.0-20200331202046-9d5940d49312 // indirect
+	google.golang.org/api v0.14.0
+	gopkg.in/ini.v1 v1.51.0
+	gopkg.in/yaml.v2 v2.2.8
+	k8s.io/api v0.17.2
+	k8s.io/apiextensions-apiserver v0.17.1
+	k8s.io/apimachinery v0.17.3
```
Comment (dgoodwin) [inline go.mod:46]: Can you clarify how you accomplished this? I'm confused by the mix of 1.17 and 1.14 versions for Kube things here, but then we seem to override to 1.14 with replace lines below?

Was there any prior art you followed to pull this off?

**Comment by dgoodwin** [comment]
Docs update can be a separate PR if we can get a clean test run here.

**Comment by dgoodwin** [comment]
Failing on:

```
2020/04/02 11:58:21 Executing test unit
go generate ./pkg/... ./cmd/...
package k8s.io/code-generator/cmd/deepcopy-gen: cannot find package "." in:
	/go/src/github.com/openshift/hive/vendor/k8s.io/code-generator/cmd/deepcopy-gen
pkg/apis/apis.go:2: running "go": exit status 1
package k8s.io/code-generator/cmd/deepcopy-gen: cannot find package "." in:
	/go/src/github.com/openshift/hive/vendor/k8s.io/code-generator/cmd/deepcopy-gen
pkg/hive/apis/hive/doc.go:2: running "go": exit status 1
make: *** [generate] Error 1
2020/04/02 11:59:04 Container test in pod unit failed, exit code 2, reason Error
2020/04/02 11:59:04 No custom metadata found and prow metadata already exists. Not updating the metadata.
2020/04/02 11:59:05 Ran for 2m50s
2020/04/02 11:59:05 Submitted failure event to sentry (id=79a2c95febc54ef3b9422ce47aca3144)
error: could not run steps: step unit failed: test "unit" failed: the pod ci-op-bm852wm2/unit failed after 32s (failed containers: test): ContainerFailed one or more containers exited
```

File: go.mod (hunk 1 above)
Comment (staebler) [reply go.mod:46]: That's a good question. The `require` stanza was entirely generated by the go tooling. That is the list of modules required by the code and its dependencies. When I tried making manual changes to that, the go tooling would overrule me with what it discovered. It was generated through a couple go commands.
```
go mod init github.com/openshfit/hive
go get -d github.com/openshfit/installer@d200b241c1844edc8609d1865bf9eeda1de95f85
go get -d github.com/openshift/cluster-network-operator@c226dcab667ea57774c81b73102f08bff5c2d518
go mod tidy
make build
make test
go mod tidy
```

But, I do see that I can make some changes. Let me play with it to see if I can get the requires to show the versions that we want.

Comment (staebler) [reply go.mod:46]: Ultimately, even after setting everything to the versions that we want in the `require` stanza, the go tooling fills it back in with other versions that it finds as upgraded versions needed by dependencies.

Comment (dgoodwin) [reply go.mod:46]: So it's essentially require automatically determining that we need things like kube 1.17, but then we override with repalcements and force it back to 1.14? I'm slightly shocked this is compiling but that's great that it is.

Comment (staebler) [reply go.mod:46]: Right.

For example,
```
$ go list -m -json k8s.io/api
{
	"Path": "k8s.io/api",
	"Version": "v0.17.2",
	"Replace": {
		"Path": "k8s.io/api",
		"Version": "v0.0.0-20190409021203-6e4e0e4f393b",
		"Time": "2019-04-09T02:12:03Z",
		"Dir": "/home/staebler/go/pkg/mod/k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b",
		"GoMod": "/home/staebler/go/pkg/mod/cache/download/k8s.io/api/@v/v0.0.0-20190409021203-6e4e0e4f393b.mod"
	},
	"Dir": "/home/staebler/go/pkg/mod/k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b",
	"GoMod": "/home/staebler/go/pkg/mod/k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b/go.mod"
}
```

Comment (staebler) [reply go.mod:46]: And it goes through all of the following to determine the latest version for the require.
```
$ go mod graph | grep "k8s.io/api@"
github.com/openshift/hive k8s.io/api@v0.17.2
github.com/openshift/machine-api-operator@v0.2.1-0.20191128180243-986b771e661d k8s.io/api@v0.0.0-20190918195907-bd6ac527cfd2
github.com/openshift/machine-api-operator@v0.2.1-0.20200310180732-c63fa2b143f0 k8s.io/api@v0.17.2
github.com/openshift/installer@v0.9.0-master.0.20200331173433-d200b241c184 k8s.io/api@v0.17.2
github.com/openshift/client-go@v0.0.0-20191216194936-57f413491e9e k8s.io/api@v0.17.0
github.com/openshift/cluster-api-provider-ovirt@v0.1.1-0.20200128081049-840376ca5c09 k8s.io/api@v0.0.0-20190918155943-95b840bb6a1f
github.com/openshift/client-go@v0.0.0-20200116152001-92a2713fa240 k8s.io/api@v0.17.1
github.com/openshift/cluster-api@v0.0.0-20191129101638-b09907ac6668 k8s.io/api@v0.0.0-20190918155943-95b840bb6a1f
github.com/openshift/machine-config-operator@v4.2.0-alpha.0.0.20190917115525-033375cbe820+incompatible k8s.io/api@v0.17.1
github.com/openshift/cluster-api-provider-libvirt@v0.2.1-0.20191219173431-2336783d4603 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
github.com/openshift/runtime-utils@v0.0.0-20191011150825-9169de69ebf6 k8s.io/api@v0.0.0-20191003035645-10e821c09743
github.com/openshift/cluster-etcd-operator@v0.0.0-alpha.0.0.20191025163650-5854b5c48ce4 k8s.io/api@v0.0.0-20190313235455-40a48860b5ab
github.com/openshift/library-go@v0.0.0-20190619114638-6b58b672ee58 k8s.io/api@v0.17.0
github.com/cri-o/cri-o@v1.9.0-beta.2.0.20191003162030-4775e1c05c26 k8s.io/api@v0.0.0
github.com/containers/buildah@v1.11.2 k8s.io/api@v0.0.0-20190813020757-36bff7324fb7
github.com/containers/libpod@v1.5.1 k8s.io/api@v0.0.0-20190624085159-95846d7ef82a
github.com/openshift/library-go@v0.0.0-20200210105614-4bf528465627 k8s.io/api@v0.17.0
github.com/openshift/cluster-api@v0.0.0-20191030113141-9a3a7bbe9258 k8s.io/api@v0.0.0-20190918155943-95b840bb6a1f
github.com/openshift/library-go@v0.0.0-20191003152030-97c62d8a2901 k8s.io/api@v0.17.0
github.com/hashicorp/vault-plugin-auth-kubernetes@v0.5.2-0.20190925162726-2e5b0b8184e6 k8s.io/api@v0.0.0-20190409092523-d687e77c8ae9
github.com/openshift/cluster-api-provider-gcp@v0.0.1-0.20200120152131-1b09fd9e7156 k8s.io/api@v0.0.0-20191003000013-35e20aa79eb8
k8s.io/kubelet@v0.17.1 k8s.io/api@v0.17.1
k8s.io/kubectl@v0.0.0-20200124035537-9f7d91504e51 k8s.io/api@v0.0.0-20200124032216-924612ff3bca
sigs.k8s.io/controller-runtime@v0.0.0-20190520212815-96b67f231945 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-tools@v0.2.2-0.20190919191502-76a25b63325a k8s.io/api@v0.0.0-20190222213804-5cb15d344471
sigs.k8s.io/controller-runtime@v0.2.0 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-runtime@v0.2.0-beta.2 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/cluster-api-provider-aws@v0.0.0 k8s.io/api@v0.17.2
sigs.k8s.io/cluster-api-provider-azure@v0.0.0 k8s.io/api@v0.0.0-20191121015604-11707872ac1c
sigs.k8s.io/controller-tools@v0.2.4 k8s.io/api@v0.0.0-20190222213804-5cb15d344471
sigs.k8s.io/controller-runtime@v0.4.0 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-runtime@v0.3.1-0.20191016212439-2df793d02076 k8s.io/api@v0.0.0-20190409021203-6e4e0e4f393b
sigs.k8s.io/controller-tools@v0.2.2-0.20190930215132-4752ed2de7d2 k8s.io/api@v0.0.0-20190222213804-5cb15d344471
```

**Comment by staebler** [comment]
> Docs update can be a separate PR if we can get a clean test run here.

Docs have been updated.

**Pull request merged**